src/domain.py

```diff
@@ -357,3 +363,111 @@ def build_prometheus_config(charm_config):
 
     logger.debug("Build prom config: {}".format(prometheus_config))
     return prometheus_config
+
+
+def _prometheus_http_api_call(
+        model_name, app_name, method, endpoint, return_response=True):
+    host = "{0}.{1}.svc:{2}".format(
+        app_name, model_name, PROMETHEUS_ADVERTISED_PORT
+    )
+    conn = http.client.HTTPConnection(host)
+    conn.request(method=method, url=endpoint)
+    logger.debug("Calling Prom API: {0} {1}".format(method, endpoint))
+
+    # TODO: Handle the un-available API endpoint case
+    response = conn.getresponse()
+    if response.status < 200 or response.status >= 300:
+        logger.error("API returned error: {0}".format(
+            response.read()
+        ))
+        raise PrometheusAPIError("Prom API returned error, see unit logs")
+
+    if return_response:
+        try:
+            return json.loads(response.read())
+        except (ValueError, TypeError) as e:
+            logger.error("Prom API returned non-JSON: {0}".format(e))
+            raise PrometheusAPIError("Non-JSON response returned")
+
+
+def reload_configuration(fw_adapter):
```

Comment:
TBH having a domain function be aware of `fw_adapter` and its intricacies feels a little iffy to me. I wanted to keep all knowledge of the framework in `charm.py` and let `domain.py` be only concerned with business logic.